[Urgent] Tray 2.1.2  Java 8 Update 291 - Print HTML Error

Hi,
QZTray 2.1.1-2.1.2
**Java 8 Update 291** (QZTray works fine with the old version but got the bugs with this version)
USB ESC/POS Thermal Receipt Printer 80mm
QZ Tray is failed in HTML when using the new Java version (8 Update 291).
Print successfully at the first time
From the second time it was printed out a very **short paragraph** with no content.

<img src="https://user-images.githubusercontent.com/17844532/113091068-556b5c80-9215-11eb-9ae8-b844da2c2211.jpg" width=300px/>

> Please try with AdoptOpenJDK 11. It will not conflict with Java 8 and uses a newer copy of the HTML renderer
> 
> You will need to restart QZ Tray after installing AdoptOpenJDK 11.
> 
> You can find out which Java version QZ is using by visiting http://localhost:8182.

I tried using AdoptOpenJDK 11 and it runs fine on Windows 64bit. But I got error on Windows 32bit, Tray can't not print any thing, this is log contents:
```
2021-03-31 18:40:52,549 [ERROR] Failed to print
java.lang.UnsupportedOperationException: Unable to start JavaFX service
```

--- a/src/qz/printer/action/WebApp.java
+++ b/src/qz/printer/action/WebApp.java
@@ -226,7 +226,7 @@ public class WebApp extends Application {
 
         // Fix blank pages for WebKit > 609.1
         // See also https://github.com/qzind/tray/issues/778
-        if(getWebkitVersion() == null || getWebkitVersion().greaterThanOrEqualTo(Version.forIntegers(609, 1, 0))) {
+        if(getWebkitVersion() == null || getWebkitVersion().greaterThan(Version.forIntegers(609, 1, 0))) {
             VECTOR_FRAMES = 30; // 30 pulses needed for vector graphics
         }
 

--- a/src/qz/printer/action/WebApp.java
+++ b/src/qz/printer/action/WebApp.java
@@ -491,13 +491,14 @@ public class WebApp extends Application {
         if(webkitVersion == null) {
             if(webView != null) {
                 String userAgent = webView.getEngine().getUserAgent();
-                String[] split = userAgent.split("WebKit/");
-                if (split.length > 1) {
-                    System.out.println("   split[1]: " + split[1]);
-                    String[] split2 = split[1].split(" ");
-                    if (split2.length > 0) {
-                        System.out.println("   split2[0]: " + split2[0]);
-                        webkitVersion = Version.valueOf(split2[0]);
+                String[] parts = userAgent.split("WebKit/");
+                if (parts.length > 1) {
+                    String[] split = parts[1].split(" ");
+                    if (split.length > 0) {
+                        try {
+                            webkitVersion = Version.valueOf(split[0]);
+                            log.info("WebKit version {} detected", webkitVersion);
+                        } catch(Exception ignore) {}
                     }
                 }
                 if(webkitVersion == null) {

--- a/src/qz/printer/action/WebApp.java
+++ b/src/qz/printer/action/WebApp.java
@@ -49,7 +49,9 @@ public class WebApp extends Application {
     private static final Logger log = LoggerFactory.getLogger(WebApp.class);
 
     private static WebApp instance = null;
-
+    private static Version webkitVersion = null;
+    private static int CAPTURE_FRAMES = 2;
+    private static int VECTOR_FRAMES = 1;
     private static Stage stage;
     private static WebView webView;
     private static double pageWidth;
@@ -221,6 +223,13 @@ public class WebApp extends Application {
         log.debug("Started JavaFX");
 
         webView = new WebView();
+
+        // Fix blank pages for WebKit > 609.1
+        // See also https://github.com/qzind/tray/issues/778
+        if(getWebkitVersion() == null || getWebkitVersion().greaterThanOrEqualTo(Version.forIntegers(609, 1, 0))) {
+            VECTOR_FRAMES = 30; // 30 pulses needed for vector graphics
+        }
+
         st.setScene(new Scene(webView));
         stage = st;
         stage.setWidth(1);
@@ -247,66 +256,67 @@ public class WebApp extends Application {
         raster = false;
 
         load(model, (int frames) -> {
-            try {
-                double printScale = 72d / 96d;
-                webView.getTransforms().add(new Scale(printScale, printScale));
+            if(frames == VECTOR_FRAMES) {
+                try {
+                    double printScale = 72d / 96d;
+                    webView.getTransforms().add(new Scale(printScale, printScale));
 
-                PageLayout layout = job.getJobSettings().getPageLayout();
-                if (model.isScaled()) {
-                    double viewWidth = webView.getWidth() * printScale;
-                    double viewHeight = webView.getHeight() * printScale;
+                    PageLayout layout = job.getJobSettings().getPageLayout();
+                    if (model.isScaled()) {
+                        double viewWidth = webView.getWidth() * printScale;
+                        double viewHeight = webView.getHeight() * printScale;
 
-                    double scale;
-                    if ((viewWidth / viewHeight) >= (layout.getPrintableWidth() / layout.getPrintableHeight())) {
-                        scale = (layout.getPrintableWidth() / viewWidth);
-                    } else {
-                        scale = (layout.getPrintableHeight() / viewHeight);
-                    }
-                    webView.getTransforms().add(new Scale(scale, scale));
-                }
-
-                Platform.runLater(() -> {
-                    double useScale = 1;
-                    for(Transform t : webView.getTransforms()) {
-                        if (t instanceof Scale) { useScale *= ((Scale)t).getX(); }
-                    }
-
-                    PageLayout page = job.getJobSettings().getPageLayout();
-                    Rectangle printBounds = new Rectangle(0, 0, page.getPrintableWidth(), page.getPrintableHeight());
-                    log.debug("Paper area: {},{}:{},{}", (int)page.getLeftMargin(), (int)page.getTopMargin(),
-                              (int)page.getPrintableWidth(), (int)page.getPrintableHeight());
-
-                    Translate activePage = new Translate();
-                    webView.getTransforms().add(activePage);
-
-                    int columnsNeed = Math.max(1, (int)Math.ceil(webView.getWidth() / printBounds.getWidth() * useScale - 0.1));
-                    int rowsNeed = Math.max(1, (int)Math.ceil(webView.getHeight() / printBounds.getHeight() * useScale - 0.1));
-                    log.debug("Document will be printed across {} pages", columnsNeed * rowsNeed);
-
-                    try {
-                        for(int row = 0; row < rowsNeed; row++) {
-                            for(int col = 0; col < columnsNeed; col++) {
-                                activePage.setX((-col * printBounds.getWidth()) / useScale);
-                                activePage.setY((-row * printBounds.getHeight()) / useScale);
-
-                                job.printPage(webView);
-                            }
-
-                            unlatch(null);
+                        double scale;
+                        if ((viewWidth / viewHeight) >= (layout.getPrintableWidth() / layout.getPrintableHeight())) {
+                            scale = (layout.getPrintableWidth() / viewWidth);
+                        } else {
+                            scale = (layout.getPrintableHeight() / viewHeight);
                         }
+                        webView.getTransforms().add(new Scale(scale, scale));
                     }
-                    catch(Exception e) {
-                        unlatch(e);
-                    }
-                    finally {
-                        //reset state
-                        webView.getTransforms().clear();
-                    }
-                });
-            }
-            catch(Exception e) { unlatch(e); }
 
-            return true; //only runs on first frame
+                    Platform.runLater(() -> {
+                        double useScale = 1;
+                        for(Transform t : webView.getTransforms()) {
+                            if (t instanceof Scale) { useScale *= ((Scale)t).getX(); }
+                        }
+
+                        PageLayout page = job.getJobSettings().getPageLayout();
+                        Rectangle printBounds = new Rectangle(0, 0, page.getPrintableWidth(), page.getPrintableHeight());
+                        log.debug("Paper area: {},{}:{},{}", (int)page.getLeftMargin(), (int)page.getTopMargin(),
+                                  (int)page.getPrintableWidth(), (int)page.getPrintableHeight());
+
+                        Translate activePage = new Translate();
+                        webView.getTransforms().add(activePage);
+
+                        int columnsNeed = Math.max(1, (int)Math.ceil(webView.getWidth() / printBounds.getWidth() * useScale - 0.1));
+                        int rowsNeed = Math.max(1, (int)Math.ceil(webView.getHeight() / printBounds.getHeight() * useScale - 0.1));
+                        log.debug("Document will be printed across {} pages", columnsNeed * rowsNeed);
+
+                        try {
+                            for(int row = 0; row < rowsNeed; row++) {
+                                for(int col = 0; col < columnsNeed; col++) {
+                                    activePage.setX((-col * printBounds.getWidth()) / useScale);
+                                    activePage.setY((-row * printBounds.getHeight()) / useScale);
+
+                                    job.printPage(webView);
+                                }
+
+                                unlatch(null);
+                            }
+                        }
+                        catch(Exception e) {
+                            unlatch(e);
+                        }
+                        finally {
+                            //reset state
+                            webView.getTransforms().clear();
+                        }
+                    });
+                }
+                catch(Exception e) { unlatch(e); }
+            }
+            return frames >= VECTOR_FRAMES;
         });
 
         log.trace("Waiting on print..");
@@ -332,7 +342,7 @@ public class WebApp extends Application {
         raster = true;
 
         load(model, (int frames) -> {
-            if (frames == 2) {
+            if (frames == CAPTURE_FRAMES) {
                 log.debug("Attempting image capture");
 
                 Toolkit.getToolkit().addPostSceneTkPulseListener(new TKPulseListener() {
@@ -354,7 +364,7 @@ public class WebApp extends Application {
                 Toolkit.getToolkit().requestNextPulse();
             }
 
-            return frames >= 2;
+            return frames >= CAPTURE_FRAMES;
         });
 
         log.trace("Waiting on capture..");
@@ -475,5 +485,28 @@ public class WebApp extends Application {
             return !graphicsJar.contains("osx") && !graphicsJar.contains("mac");
         }
         return !graphicsJar.contains("linux");
+    }
+
+    public static Version getWebkitVersion() {
+        if(webkitVersion == null) {
+            if(webView != null) {
+                String userAgent = webView.getEngine().getUserAgent();
+                String[] split = userAgent.split("WebKit/");
+                if (split.length > 1) {
+                    System.out.println("   split[1]: " + split[1]);
+                    String[] split2 = split[1].split(" ");
+                    if (split2.length > 0) {
+                        System.out.println("   split2[0]: " + split2[0]);
+                        webkitVersion = Version.valueOf(split2[0]);
+                    }
+                }
+                if(webkitVersion == null) {
+                    log.warn("WebKit version couldn't be parsed from UserAgent: {}", userAgent);
+                }
+            } else {
+                log.warn("Can't get WebKit version, JavaFX hasn't started yet.");
+            }
+        }
+        return webkitVersion;
     }
 }

--- a/src/qz/utils/SystemUtilities.java
+++ b/src/qz/utils/SystemUtilities.java
@@ -501,7 +501,7 @@ public class SystemUtilities {
             try {
                 Class.forName("com.sun.glass.ui.monocle.MonoclePlatformFactory");
                 hasMonocle = true;
-            } catch (ClassNotFoundException e) {
+            } catch (ClassNotFoundException | UnsupportedClassVersionError e) {
                 hasMonocle = false;
             }
         }

